[Wasm] Add support for baseUrl and mainScriptPath

jeromelaban opened the pull request:
This change allows for resources like the crypto worker or thread workers to be resolved properly when used in a generic context.

pavelsavara commented:
See also this https://github.com/dotnet/runtime/pull/69441

kg commented on src/mono/wasm/runtime/crypto-worker.ts:
```diff
@@ -41,7 +41,7 @@ export function init_crypto(): void {
 
         const chan = LibraryChannel.create(1024); // 1024 is the buffer size in char units.
-        const worker = new Worker("dotnet-crypto-worker.js");
+        const worker = new Worker(`${Module.baseUrl}dotnet-crypto-worker.js`);
```
if Module.baseUrl is undefined this would cause path resolution to fail in a strange way by inserting 'undefined' at the front, it would be better to default it to '' or check and throw. In both cases the situation is worse because errors in web workers are typically swallowed (it's possible an invalid worker path would throw in the browser tab, at least)

kg replied:
nvm, I see you take steps to make sure it's initialized, I missed it

kg commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -36,18 +36,33 @@ export function configure_emscripten_startup(module: DotnetModule, exportedAPI:
+            }
+
             (<any>module)["mainScriptUrlOrBlob"] = hr;
         } else {
```
while we're changing this, it might make sense to have a convention for how to tell the runtime what its baseUrl is so it doesn't have to guess. i'm guessing uno could pipe that through

lambdageek replied:
The guessing should be removed - and blazor updated to provide a base URL (and possibly also the URL of the worker script - actually depending on how much stuff they mangle we might need some locate function where we can ask for normal names that we know and blazor can tell us what they mangled them to).  the guessing was just a hack because I didn't want to coordinate a pair of aspnetcore and runtime PRs

jeromelaban committed:
feat: Add support for baseUrl and mainScriptPath

This change allows for resources like the crypto worker or thread workers to be resolved properly when used in a generic context.

pavelsavara commented:
guys, I don't have bandwidth to fully explore this space right now, but please make sure that you align it with NodeJS and that we have some consistent plan for the relative paths of configured resources. We have open discussion on the PR I mentioned above and also we have @radical 's host and maybe some new shape of the config.

All that said, I think we should just add `dotnet-crypto-worker.js` into `mono-config.json` as new resource type, rather than hard-code it here.

pavelsavara commented:
Crypto worker is gone now. 
For threads worker we do `resolve_asset_path` which uses now improved `locateFile`.

https://github.com/dotnet/runtime/blob/3c21cc5a0a486d08b5a7a4fdc2e1c2f9b02dbc9f/src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts#L39-L42

So, I guess this PR is obsolete ? @jeromelaban

jeromelaban commented:
@pavelsavara I guess it is! Thanks for the update.
